### ISSUE TITLE
Char CNN + BiLSTM

### ISSUE THREAD

**Issue by aleksei-ai** [deeplearning4j/deeplearning4j#4513]
#### Issue Description

I want to create a Neural Net for solving NER task from CoNLL 2003 https://www.clips.uantwerpen.be/conll2003/ner/

I want to implement Char CNN - BiLSTM architecture
![image](https://user-images.githubusercontent.com/31881382/35037901-f4a69b14-fb89-11e7-8c28-08a0cb1310cb.png)

I managed to create Word Embeddings + BiLSTM network.

So now I need:
1. Lookup for every character to Char Embeddings
2. Make convolutions for every word
3. Make max pooling for every word
4. Concatenate this char-based word representation with word embeddings

Here is my GIST https://gist.github.com/aleksei-ai/7200e80634611a3adbb8efaaed6160ce

Can you suggest how to do it?
- How organize char in Iterator?
- How to concatenate input char representation and word embeddings?

#### Environment
* Deeplearning4j version - 0.9.1

**Comment by tomthetrainer** [comment]
@eraly Might have some input, Not that we will write the code for you as you seem capable, but perhaps some direction or guidance.

**Comment by aleksei-ai** [comment]
@tomthetrainer that would be great!

**Comment by aleksei-ai** [comment]
@tomthetrainer Anything new here?

**Comment by tomthetrainer** [comment]
Let me ask for an engineer to take a look

**Comment by AlexDBlack** [comment]
So, there's nothing out of the box in DL4J that will do exactly this. You're unfortunately looking at a custom DataSetIterator.

It's not exactly clear to me what you want to achieve here. Different possible interpretations of your architecture will have different data pipeline and network designs. Specifically:
1. Are the word embeddings fixed/static (like w2v vectors), or are they trained along with the rest of  the net?
2. As for the "CNN-extracted Char features" - again, are these fixed (i.e.., pulled from some separately trained network) or are they trained in the same network as the RNN layer(s)?

Beyond that: on the off chance that if you haven't seem them already, CnnSentenceDataSetIterator (in DL4J repo) and the Character iterator (examples repo) are probably worth a look.

**Comment by aleksei-ai** [comment]
@AlexDBlack 
1. I want to have an opportunity to Initialize word embeddings (from word2vec or glove). But then I want to finetune it during training. Is it possible?
2. It's trained in the same network. Just before RNN (you can see on the picture above).

I looked at CnnSentenceDataSetIterator. It's not very useful for me.
Also I found way to build architecture. I have to use several ReshapeVertex and PreprocessorVertex. Will describe here my results. 

Also if you have any thoughts it could be useful for me.

**Comment by AlexDBlack** [comment]
Initialize with w2v/glove then fine tune: yes, totally possible. At present, you need to manually copy them over from the w2v parameters to the EmbeddingLayer. Something like net.getLayer("embeddingLayer").getParam("W").assign(<word vector params>) should be sufficient.

Ok, so CNN is trained in the net. That's possible.

OK, so custom MultiDataSetIterator then... you'll have 3 input features:
1. Word indexes: INDArray with shape [minibatch, 1, timeSeriesLength]. Each value is as integer in range [0, vocabSize-1]. Feeds in to the embedding layer, which then feeds in to the bidirectional LSTM
2. Additional word features: INDArray with shape [minibatch, featureSize, timeSeriesLength]. i.e., same as a simple/normal RNN. Feeds in directly to the bidirectional LSTM.
3. CNN extracted char features: this is the more complex part.

Where "timeSeriesLength" is the length of the longest sentence in the minibatch.

One issue here is that obviously each word can have an arbitrary number of characters, but ultimately we need to have a fixed length vector. From what I understand, you approach is to do timeDistributed(CNN -> global max pooling) -> RNN.
DL4J doesn't have fantastic time distributed CNN functionality, though we can do it.
Features here: [minibatch, X, timeSeriesLength] where ```X = numChars * maxWordLength```
Basically for each time step, we start with a [minibatch, depth, height, width] = [1, 1, maxWordLength, numChars] and do a c-order reshape to give us the ```maxWordLength * numChars``` vector.
RnnToCnnPreProcessor will then handle the reshaping (time distributed) part to 'undo' the flattening ready for the CNN layers.
I hope that makes some sense.


Your labels array is pretty standard for a RNN: [minibatch, numClasses, timeSeriesLength]. You'll also need label masks here.

**Comment by aleksei-ai** [comment]
@AlexDBlack Thank you for response.

I've create this architecture https://gist.github.com/aleksei-ai/7200e80634611a3adbb8efaaed6160ce
But now I got A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f4423934ec5, pid=26698, tid=0x00007f44244a9700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libc.so.6+0x84ec5]# [ timer expired, abort... ]

I'm using Ubuntu 17.04

**Comment by treo** [comment]
@aleksei-ai what you are seeing there is a JVM crash... can you create a new issue along with the crash log please?

**Comment by aleksei-ai** [comment]
@AlexDBlack I switched to 0.9.2-SNAPSHOT.
Probably there is a bug in Bidirectional. 
This code throws NullPointerException: 
    ```
.addLayer("sentence_repr", new Bidirectional.Builder(Mode.CONCAT,
      new LSTM.Builder()
        .name("sentence_lstm")
        .nIn(wordDims + charDims)
        .nOut(lstmLayerSize)
        .activation(Activation.TANH)
        .weightInit(WeightInit.XAVIER)
        .build()
    )
      .dropOut(0.5)
      .build(), "word_repr")
```
As I understand the problem is that private constructor     
```
   private Bidirectional(Bidirectional.Builder builder) {
        super(builder);
    }

```is not initialize variables
```
    private Layer fwd;
    private Layer bwd;
    private Mode mode;
```

**Comment by aleksei-ai** [comment]
@AlexDBlack 
Also as I see in 0.9.2-SNAPSHOT there are breaking changes.
So may I ask you several questions how to use it:
- How to set Multi Class Cross Entropy in RnnOutputLayer? In 0.9.1 I use Enum for it
- How to set Learning Rate? I don't see this method in builder in 0.9.2-SNAPSHOT

**Comment by AlexDBlack** [comment]
@aleksei-ai there are no changes to how MCXENT is set on RnnOutputLayer between 0.9.1 and current master. You should still be able to use the enumeration config as before (let me know if this isn't the case).

As for the learning rate: it's part of the updater only now (so: .updater(new Adam(learningRate)) etc).
Why? For schedules and the fact that not all updaters have learning rates...

**Comment by AlexDBlack** [comment]
Closing due to inactivity.
